presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/IndexWriter.java

```diff
@@ -0,0 +1,353 @@
+    }
+
+    public static void createIndexTable(Database.Type databaseType, Handle handle, long tableId, List<ColumnInfo> columns, OptionalLong temporalColumnId)
+    {
+        StringJoiner tableColumns = new StringJoiner(",\n  ", "  ", ",\n").setEmptyValue("");
+
+        for (ColumnInfo column : columns) {
+            String columnType = sqlColumnType(databaseType, column.getType());
+            if (columnType != null) {
+                tableColumns.add(format("%s %s", minColumn(column.getColumnId()), columnType));
+                tableColumns.add(format("%s %s", maxColumn(column.getColumnId()), columnType));
+            }
+        }
+
+        String coveringIndex = "";
+        if (temporalColumnId.isPresent()) {
+            // add the max temporal column first to accelerate queries that usually scan recent data
+            coveringIndex = format(
+                    "  UNIQUE (%s, %s, bucket_number, chunk_id, start_commit_id, end_commit_id),\n",
+                    minColumn(temporalColumnId.getAsLong()),
+                    maxColumn(temporalColumnId.getAsLong()));
+        }
+
+        String createTable = "" +
+                "CREATE TABLE %s (\n" +
+                "  start_commit_id BIGINT NOT NULL,\n" +
+                "  end_commit_id BIGINT,\n" +
+                "  chunk_id BIGINT NOT NULL,\n" +
+                "  bucket_number INT NOT NULL,\n" +
+                "%s" +
+                "  PRIMARY KEY (bucket_number, chunk_id),\n" +
+                "%s" +
+                "  UNIQUE (chunk_id),\n" +
+                "  UNIQUE (start_commit_id, chunk_id),\n" +
+                "  UNIQUE (end_commit_id, chunk_id)\n" +
+                ")";
+        String sql = format(createTable, chunkIndexTable(tableId), tableColumns, coveringIndex);
+
+        handle.execute(sql);
+    }
+
+    public static void addIndexTableColumn(Database.Type databaseType, Handle handle, long tableId, ColumnInfo column)
+    {
+        String columnType = sqlColumnType(databaseType, column.getType());
+        if (columnType == null) {
+            return;
+        }
+
+        String tableName = chunkIndexTable(tableId);
+        String minColumn = format("%s %s", minColumn(column.getColumnId()), columnType);
+        String maxColumn = format("%s %s", maxColumn(column.getColumnId()), columnType);
+
+        String sql;
+        if (databaseType == Database.Type.H2) {
+            sql = format("ALTER TABLE %s ADD COLUMN (%s, %s)", tableName, minColumn, maxColumn);
+        }
+        else {
+            sql = format("ALTER TABLE %s ADD COLUMN %s, ADD COLUMN %s", tableName, minColumn, maxColumn);
+        }
+
+        handle.execute(sql);
+    }
+
+    public static void dropIndexTable(Handle handle, long tableId)
+    {
+        handle.execute("DROP TABLE IF EXISTS " + chunkIndexTable(tableId));
+    }
+
+    public static void dropIndexTableColumn(Database.Type databaseType, Handle handle, long tableId, long columnId)
```

Comment:
I only find dropIndexTableColumn() called by DatabaseMetadataWriter.recover. So index table can have more columns than we expect.
Maybe we should call it in more places. DatabaseMetadataWriter.dropColumn may not be the place, since index table isn't MVCC, some transaction still writing into index table may suddenly fail.

Reply:
Good catch, we should clean these up in `CommitCleaner`. This is safe to do since at the point we are cleaning up a commit, there are no active commits that reference the version of the table with the column ID that was dropped.